feat: Github PR creation

Xaroz opened the pull request:
This PR allows the deploy app to offer auto creation of pull request to the registry using `octokit` package as a base.
Fixes [ENG-1671](https://linear.app/hyperlane-xyz/issue/ENG-1671/offer-auto-opening-pr-to-registry)

- Creates handler for an API service where users can submit their deploy configs
- The service is fully validated before being able to submit
- Creates Modal for users to accept submitting a PR

TODO:

- [x] Add changeset
- [x] Run eslint rules properly on configs 
- [x] Add validation/auth to avoid repeated requests/API abuse

For the Eslint rules instead of using ESLint which is not allowed in the browser, I sorted the tokens by chain name and the connections by their chain name too, which would match the rules we have in the registry plus yaml takes care of sorting the config by their keys


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added the ability to create GitHub pull requests directly from the app for adding Warp route deployments to the Hyperlane Registry, including form validation, logo uploads, and signature verification.
  * Introduced a modal interface for submitting registry pull requests with user-friendly input fields and feedback.
  * Added new utility functions for sorting configuration data and normalizing input.
  * Provided a new image input component for file uploads.
  * Added server-side support for GitHub API client caching and PR creation with signature verification.
  * Enhanced deployment success screen with Warp route PR submission flow and improved config handling.
  * Added a new API endpoint to handle pull request creation with validation and signature verification.

* **Bug Fixes**
  * Improved error message extraction in toast notifications for better clarity.

* **Chores**
  * Updated environment variable examples and TypeScript configuration.
  * Added and updated dependencies for GitHub integration and ID generation.
  * Enhanced Next.js configuration with experimental optimizations and improved security headers.
  * Refined ESLint configuration key formatting.

* **Documentation**
  * Extended environment variable documentation and improved code comments for clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Xaroz committed:
feat: initial create PR setup

Xaroz committed:
chore: move to useMutation

Xaroz committed:
chore: clean up

Xaroz committed:
chore: improve typing and symbol check

Xaroz committed:
chore: validate configs, clean up pr descriptions

Xaroz committed:
chore: include modal for adding PR, fix bug with modal close

Xaroz committed:
feat: include username, org and clean up

Xaroz committed:
chore: move PR creation to success, clean up hard-coded index and landing

Xaroz committed:
feat: add changeset to commits

Xaroz committed:
chore: sort warp core config

xeno097 commented on src/features/deployment/github.ts:
```diff
@@ -0,0 +1,77 @@
+): Promise<CreatePrResponse> {
+  const deployConfigFilename = getDeployConfigFilename(deployConfig);
+  const warpConfigFilename = getWarpConfigFilename(warpConfig);
+  const firstNonSynthetic = Object.values(deployConfig).find((c) => !isSyntheticTokenType(c.type));
+
+  if (!firstNonSynthetic) throw new Error('Token types cannot all be synthetic');
```
I think that you can use the WarpDeploySchema from the sdk to have this same misconfig checking

xeno097 commented on src/features/deployment/github.ts:
```diff
@@ -0,0 +1,77 @@
+  const warpConfigFilename = getWarpConfigFilename(warpConfig);
+  const firstNonSynthetic = Object.values(deployConfig).find((c) => !isSyntheticTokenType(c.type));
+
+  if (!firstNonSynthetic) throw new Error('Token types cannot all be synthetic');
```
I think that you can use the WarpDeploySchema from the sdk to have this same misconfig checking

Xaroz replied:
I can do this in another PR since the warp route form uses this and wanna make sure I am not messing up anything

xeno097 commented on src/pages/api/create-pr.ts:
```diff
@@ -0,0 +1,123 @@
+    return res.status(500).json({
+      error: 'Missing Github configurations, check your environment variables',
+    });
+
```
Why not create a util function that helps you in creating API responses so you can hide the `res.status(...).json(...)` calls?

xeno097 commented on src/pages/api/create-pr.ts:
```diff
@@ -0,0 +1,123 @@
+    GithubIdentity & { symbol: string };
+
+  if (!deployConfig || !warpConfig || !symbol)
+    return res.status(400).json({ error: 'Missing config files to create PR' });
```
I would create a zod validator and validate the request, this way you don't even have to type assert because you can use the valdiated value after parsing, dunno if APIs writeen in next allow it but i would add this validation into a middlewre that gets invoked before the actual endpoint handler

Xaroz replied:
I added this into its own validation function, middleware itself behaves a bit differently in NextJS (it does not contain request.body)

xeno097 commented on src/pages/api/create-pr.ts:
```diff
@@ -0,0 +1,123 @@
+
+  const warpConfigResult = validateStringToZodSchema(warpConfig.content, WarpCoreConfigSchema);
+  if (!warpConfigResult) return res.status(400).json({ error: 'Invalid warp config' });
+
```
This should be part of the middleware too

xeno097 commented on src/pages/api/create-pr.ts:
```diff
@@ -0,0 +1,123 @@
+
+  const { githubBaseBranch, githubForkOwner, githubRepoName, githubToken, githubUpstreamOwner } =
+    serverConfigParseResult.data;
+  const octokit = new Octokit({ auth: githubToken });
```
Is there a way to get this to persist instead instead of having to create it each time the endpoint is hit?

xeno097 commented on src/pages/api/create-pr.ts:
```diff
@@ -0,0 +1,123 @@
+      sha: latestCommitSha,
+    });
+
+    const changesetFile = writeChangeset(`Add ${symbol} deploy artifacts`);
```
I'd be a bit more explcit and set the msg as `Add ${symbol} warp route deploy artifacs` and if possible i's use the warp id instead of just the symbol

Xaroz committed:
chore: sort result in deployment details

xeno097 commented on src/features/deployment/utils.ts:
```diff
@@ -23,3 +25,38 @@ export function downloadYamlFile(config: unknown | undefined, filename: string)
   URL.revokeObjectURL(url);
 }
+
+export function getDeployConfigFilename(config: WarpRouteDeployConfig) {
+  if (!config) return 'deploy.yaml';
+  const chains = objKeys(config).sort();
+
+  return `${chains.join('-')}-deploy.yaml`;
+}
+
+export function getWarpConfigFilename(config: WarpCoreConfig) {
+  if (!config) return 'config.yaml';
+
+  const chains = config.tokens.map((token) => token.chainName).sort();
+  return `${chains.join('-')}-config.yaml`;
+}
+
+export function sortWarpCoreConfig(warpCoreConfig?: WarpCoreConfig): WarpCoreConfig | undefined {
+  if (!warpCoreConfig) return undefined;
+
+  const tokens = warpCoreConfig.tokens;
+
+  const sortedTokens = [...tokens]
+    .sort((a, b) => a.chainName.localeCompare(b.chainName))
+    .map((token) => ({
+      ...token,
+      connections: token.connections
+        ? [...token.connections].sort((a, b) => {
+            const chainA = a.token.split('|')[1] || '';
+            const chainB = b.token.split('|')[1] || '';
+            return chainA.localeCompare(chainB);
+          })
+        : undefined,
+    }));
+
+  return { ...warpCoreConfig, tokens: sortedTokens };
```
I think, if  possible, it would be better to reuse the utils for generating warp ids and related file names from the registry package.
refs:
- https://github.com/hyperlane-xyz/hyperlane-registry/blob/main/src/registry/BaseRegistry.ts#L72
- https://github.com/hyperlane-xyz/hyperlane-registry/blob/main/src/registry/warp-utils.ts

Xaroz replied:
Could we move the logic inside `warpRouteConfigToId` to an utils function? That way the deploy app does not need to use an instance of the `BaseRegistry` because it has no need to

Xaroz committed:
chore: persist octokit instance, refactor json response and type

Xaroz committed:
chore: include warpRouteId to request

Xaroz committed:
chore: move create PR types to its own file

Xaroz committed:
chore: refactor body validation into its own fn

Xaroz committed:
chore: move createPr types

Xaroz committed:
chore: use warpRouteId for filenames

xeno097 commented on src/features/deployment/utils.ts:
```diff
@@ -26,18 +25,13 @@ export function downloadYamlFile(config: unknown | undefined, filename: string)
-  if (!config) return 'deploy.yaml';
-  const chains = objKeys(config).sort();
+export function getConfigsFilename(warpRouteId: string) {
+  const [_, label] = warpRouteId.split('/');
```
i don't remember if it is exported from the registry package but there is a function that allows you to split the symbol and label from a warp route id

Xaroz replied:
there is indeed a fn in the registry but it is currently not exported

xeno097 commented on src/utils/api.ts:
```diff
@@ -0,0 +1,10 @@
+  statusCode: number,
+  body?: ApiResponseBody<T>,
+) {
+  return res.status(statusCode).json(body as ApiResponseBody<T>);
```
Is the type assertion needed? if the compiler complained about body being possible undefined maybe it would be better to set `.json()` only when body is defined no?

Xaroz committed:
chore: remove type assertion

Xaroz committed:
feat: verify create-pr request to prevent spamming (#31)

fixes
[ENG-1824](https://linear.app/hyperlane-xyz/issue/ENG-1824/rate-limiting-avoid-spamming-of-endpoints)

This PR features different ways to prevent the `create-pr` endpoint from
being spammed, the process goes like:

1. In the UI, the temporary deployer wallet will sign a message that
contains a `timestamp` and `warpRouteId`.
2. The returned `signature` will be sent to the endpoint along with the
`address` and `message`
3. The backend service will verify the signature using viem's
`verifySignature` and also check if the `timestamp` has expired (its
currently set to 2 minutes)
4. A `keccak256` hash will be created using the `warpRouteId`,
`deployConfig` and `warpConfig`, this will be used for the branch name,
which will be in `warpRouteId-keccak256hash` format
5. This branch name will be verified using `octokit` so no other
branches exists with the same name
6. If every check pass then the PR is created

Rate limiting will be implemented at a later PR

<!-- This is an auto-generated comment: release notes by coderabbit.ai
-->

## Summary by CodeRabbit

* **New Features**
* Added a clear message in the PR creation modal informing users that
wallet signature is required for verification, with no on-chain
transaction involved.
* Introduced wallet-based signature verification for PR creation,
enhancing security and authenticity.
* Implemented deterministic branch naming to prevent duplicate PRs for
the same configuration.

* **Bug Fixes**
* Prevented duplicate pull requests by checking for existing branches
with the same configuration.

* **Refactor**
* Improved internal structure for PR creation and signature verification
processes to enhance maintainability.

* **Chores**
* Added utility for sorting object keys to ensure consistent
configuration handling.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Xaroz commented on src/pages/api/create-pr.ts:
```diff
@@ -0,0 +1,276 @@
+
+  const { address, message, signature } = parsedSignatureBody.data;
+
+  if (!isHex(address) || !isValidAddressEvm(address))
+    return { error: 'Address is not a valid EVM hex string' };
+  if (!isHex(signature) || !isValidAddressEvm(address))
+    return { error: 'Signature is a not a valid EVM hex string' };
+
```
we do want this check

Xaroz commented on src/features/deployment/github.ts:
```diff
@@ -0,0 +1,129 @@
+          message,
+          signature,
+        },
+      });
```
ah good job on this one!

Xaroz commented on src/consts/config.server.ts:
```diff
@@ -0,0 +1,46 @@
+
+export type ServerConfig = z.infer<typeof ServerConfigSchema>;
+
+const githubForkOwner = process.env.GITHUB_FORK_OWNER || '';
+const githubRepoName = process.env.GITHUB_REPO || '';
+const githubUpstreamOwner = process.env.GITHUB_UPSTREAM_OWNER || '';
+const githubBaseBranch = 'main';
+const githubToken = process.env.GITHUB_TOKEN || '';
+const serverEnvironment = process.env.SERVER_ENVIRONMENT || 'production';
+
+export const serverConfig: ServerConfig = Object.freeze({
+  githubForkOwner,
+  githubRepoName,
+  githubUpstreamOwner,
+  githubBaseBranch,
+  githubToken,
+  serverEnvironment,
+});
```
its being validated elsewhere using `safeParse`

Xaroz committed:
fix: wrong address for verification and assertion for warpRouteId

Xaroz committed:
chore: change flow redirection

Xaroz committed:
chore: use viem for kekkach256

Xaroz committed:
fix: attempt to fix build by excluding rainbow kit

Xaroz committed:
chore: attempt fix with standalone flag

Xaroz committed:
chore: temporary remove addressEvmCheck

Xaroz committed:
chore: return redirect flow

Xaroz committed:
chore: optimize imports

Xaroz committed:
feat: upload img for PR creation (#32)

This PR introduces a way to allow users to upload an image along with
the automated PR creation. fixes
[ENG-1825](https://linear.app/hyperlane-xyz/issue/ENG-1825/allow-users-to-upload-logos-to-automated-pr)

- Change api endpoint to use app router instead of page router, this
will give us access to multi-form data without relying on a third party
package
- Switch request body to be multi-form data instead
- New component for image uploading called `ImageInput`
- Include new field called `logo` with proper validation
- Uploads the logo to github along with the other files

Xaroz committed:
chore: better error handling

Xaroz commented on src/app/api/create-pr/route.ts:
```diff
@@ -0,0 +1,346 @@
+
+  const logoFile = await getLogoFile(logo, warpRouteId);
+
+  const sortedDeployConfig = sortObjByKeys(deployConfigResult);
+  const sortedWarpCoreConfig = sortObjByKeys(
+    sortWarpCoreConfig({
+      ...warpConfigResult,
+      tokens: warpConfigResult.tokens.map((token) => ({ ...token, logoURI: logoFile?.path })),
+    })!,
+  );
```
This shouldn't be an issue since the line above, validates the config, but I am asserting because typescript complains
```
  const warpConfigResult = validateStringToZodSchema(warpConfig.content, WarpCoreConfigSchema);
  if (!warpConfigResult) return { error: 'Invalid warp config content' };
```

Xaroz committed:
Merge branch 'main' into xaroz/feat/pr-creation
